### PR TITLE
doc: Add documentation for nRF5340

### PR DESCRIPTION
**Pull request by ru-fu**
[Working with nRF5340 — nRF Connect SDK 1.0.pdf](https://github.com/NordicPlayground/fw-nrfconnect-nrf/files/3832244/Working.with.nRF5340.nRF.Connect.SDK.1.0.pdf)

Add a user guide that explains how to work with nRF5340,
what samples are available, and how they are programmed.
Add nRF5340 DK to the list of supported boards in the
samples.
Update other documentation as required to include nRF5340.

Signed-off-by: Ruth Fuchss <ruth.fuchss@nordicsemi.no>

### CODE REVIEW

File: doc/nrf/ug_nrf5340.rst [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+The board name for the application core in Zephyr is ``nrf5340_dk_nrf5340_cpuapp``.
+
+.. note::
+   The M33 TrustZone divides the application core into secure (``nrf5340_dk_nrf5340_cpuapp``) and non-secure (``nrf5340_dk_nrf5340_cpuappns``) domains.
+   However, all nRF5340 samples currently run in the secure domain, so you should not use the ``nrf5340_dk_nrf5340_cpuappns`` board.
+
+Inter-core communication
+========================
+
+.. TODO - baso: please check
```
Comment (ru-fu) [inline doc/nrf/ug_nrf5340.rst]: @barsok

File: doc/nrf/ug_nrf5340.rst [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+The |NCS| provides a series of :ref:`Bluetooth Low Energy samples <ble_samples>`, in addition to the :ref:`Bluetooth samples in Zephyr <zephyr:bluetooth-samples>`.
+Most of these samples should run on the nRF5340 PDK, but not all have been thoroughly tested.
+Samples that use non-standard features of the Bluetooth Low Energy controller, like the :ref:`ble_llpm` sample, are not supported.
+
+.. Some samples require configuration adjustments to the :ref:`zephyr:bluetooth-hci-openamp-sample` sample as described in the `Network sample`_ section.
+
+Some samples require configuration adjustments to the network sample as described in the `Network sample`_ section.
+
+.. TODO - kapi: what adjustments are needed to the throughput sample?
```
Comment (ru-fu) [inline doc/nrf/ug_nrf5340.rst]: @kapi-no

File: samples/bluetooth/throughput/README.rst [hunk 3]
```diff
@@ -67,6 +67,9 @@ Requirements
 
 * Two of the following development boards:
 
+  .. TODO - add note about extra configuration required for nRF5340
```
Comment (ru-fu) [inline samples/bluetooth/throughput/README.rst]: @kapi-no

File: doc/nrf/ug_nrf5340.rst [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+#. Double-click the value for **Target Device** again and enter **nRF5340** in the search field.
+   Click **OK** to use this value.
+
+You can now build and program the sample for nRF5340.
+
+
+Programming from the command line
+=================================
+
+To program a HEX file after building it with |SES|, open a command prompt in the build folder of the sample that you want to program and enter the following command::
```
Comment (greg-fer) [inline doc/nrf/ug_nrf5340.rst]: This and the next paragraph could be merged into one to save some space, for example:
```
To program a HEX file after building it with |SES|, open a command prompt in the build folder of the sample that you want to program and enter the following commands depending on the method of your choice:
- No nrfjprog: `west flash`
- nrfjprog (part of the `nRF Command Line Tools`_): <commands>
```

File: doc/nrf/ug_nrf5340.rst [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+    west flash
+
+If you prefer to use nrfjprog (which is part of the `nRF Command Line Tools`_) instead, open a command prompt in the build folder of the network sample and enter the following commands to program the network sample::
```
Comment (greg-fer) [inline doc/nrf/ug_nrf5340.rst]: Is the folder for building the same as in line 141? If not, I would add italics around the first `network` to make it stand out that it's a different folder. If yes, I would rephrase for consistency.

File: samples/bluetooth/central_bas/README.rst [hunk 6]
```diff
@@ -20,7 +20,8 @@ Requirements
 
 * One of the following development boards:
 
-  * nRF9160 DK board (PCA10090)
+  * |nRF9160DK|
+  * |nRF5340DK|
   * nRF52840 Development Kit board (PCA10056)
```
Comment (greg-fer) [inline samples/bluetooth/central_bas/README.rst]: Would it make sense to add these other DKs to shortcuts as well? It would make it easier to track them in all the docs.

File: doc/nrf/ug_nrf5340.rst [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+.. :ref:`zephyr:nrf5340_dk_nrf5340` gives an overview of the nRF5340 PDK support in Zephyr.
+
+Network core
+============
+
+The network core is an Arm Cortex-M33 processor with a reduced feature set, designed for ultra-low power operation.
+
+This core is used to communicate with the network.
+With regards to the nRF5340 samples, this means that the network core runs the lower part of the radio stack (for example, the Bluetooth Low Energy controller).
+In general, this core should be used for real-time processing tasks involving low-level protocols and layers.
```
[the remark this answers is not in the document]
Comment (ru-fu) [reply doc/nrf/ug_nrf5340.rst]: No, I don't think so. This is not a complete list, but just supposed to give you an example of what you should use this core for.

File: doc/nrf/ug_nrf5340.rst [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+You might need to adjust the Kconfig configuration of this sample to make it compatible with the peer application.
+For example, :option:`zephyr:CONFIG_BT_MAX_CONN` must be equal to the maximum number of connections supported by the application sample.
+
+
+Application samples
+===================
+
+The |NCS| provides a series of :ref:`Bluetooth Low Energy samples <ble_samples>`, in addition to the :ref:`Bluetooth samples in Zephyr <zephyr:bluetooth-samples>`.
+Most of these samples should run on the nRF5340 PDK, but not all have been thoroughly tested.
```
[the remark this answers is not in the document]
Comment (ru-fu) [reply doc/nrf/ug_nrf5340.rst]: I see your point, but I can't think of a way to do it like this without stating that all samples are supported (except for the ones in line 88), and then having the note contradicting that statement.

File: doc/nrf/ug_nrf5340.rst [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+.. important::
+   When programming the samples from |SES|, you might get an error message stating that the target cannot be identified.
+   In this case, you can either modify your |SES| installation and projects to add support for programming nRF5340, or program the generated HEX files from the command line instead.
+
+   See the following sections for more information.
+
+
+Adding support for programming nRF5340 in |SES|
+===============================================
+
```
[the remark this answers is not in the document]
Comment (ru-fu) [reply doc/nrf/ug_nrf5340.rst]: You don't need this for building, only for programming. :(
But it's a temporary section anyway, so it'll be removed soon.

File: doc/nrf/ug_nrf5340.rst [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+The current version of the Nordic Edition of |SES| (v4.20a) does not include the latest SEGGER J-Link version (v6.54c).
```
[the remark this answers is not in the document]
Comment (ru-fu) [reply doc/nrf/ug_nrf5340.rst]: Removed "latest". I didn't reword as suggested because you don't know what J-Link version SES is using.

File: doc/nrf/ug_nrf5340.rst (hunk 4 above)
Comment (ru-fu) [reply doc/nrf/ug_nrf5340.rst]: No, this just makes it harder to understand imo. Most people will probably use west and can stop after reading the first paragraph.

File: doc/nrf/ug_nrf5340.rst (hunk 5 above)
Comment (ru-fu) [reply doc/nrf/ug_nrf5340.rst]: Line 141 applies to building either sample.
The instructions for nrfjprog differ depending on which sample you build.

File: samples/bluetooth/central_bas/README.rst (hunk 6 above)
Comment (ru-fu) [reply samples/bluetooth/central_bas/README.rst]: Yes, but this is not in the scope for this PR. We can do it in NCSDK-3600.

File: doc/nrf/ug_nrf5340.rst (hunk 5 above)
Comment (greg-fer) [reply doc/nrf/ug_nrf5340.rst]: I see. I think this part would benefit from re-writing anyway, as the two paragraphs look like alternatives, because of this "If... instead".

**Comment by kapi-no** [comment]
CC: @grochu, @KAGA164

File: doc/nrf/ug_nrf5340.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (greg-fer) [reply doc/nrf/ug_nrf5340.rst]: "programming" is mentioned in the first paragraph, so it clarifies the section title. Alternatively, maybe just use "SES" instead of "|SES|" in the title to make it shorter?

File: doc/nrf/ug_nrf5340.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (greg-fer) [reply doc/nrf/ug_nrf5340.rst]: OK.

**Comment by KAGA164** [comment]
Looks good for me.

**Comment by kapi-no** [comment]
@ru-fu, I covered my TODOs

File: doc/nrf/ug_nrf5340.rst [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+.. :ref:`zephyr:nrf5340_dk_nrf5340` gives an overview of the nRF5340 PDK support in Zephyr.
+
+Network core
+============
+
+The network core is an Arm Cortex-M33 processor with a reduced feature set, designed for ultra-low power operation.
+
+This core is used to communicate with the network.
+With regards to the nRF5340 samples, this means that the network core runs the lower part of the radio stack (for example, the Bluetooth Low Energy controller).
```
[the remark this answers is not in the document]
Comment (joerchan) [reply doc/nrf/ug_nrf5340.rst]: Good point. Could write it as:
" for example the controller part of the Bluetooth stack can run on the network core while the host part of the Bluetooth stack can run on the application core."

File: doc/nrf/ug_nrf5340.rst [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+
+Adding support for programming nRF5340 in SES
+=============================================
+
+The current version of the Nordic Edition of |SES| (v4.20a) does not include SEGGER J-Link version v6.54c.
+However, this J-Link version is required to program nRF5340 devices.
+
+To add support for programming nRF5340 in |SES|, complete the following steps:
+
+1. Download and install the latest `J-Link Software and Documentation Pack`_ (v6.54c or later).
```
Comment (kapi-no) [inline doc/nrf/ug_nrf5340.rst]: SES v4.30a uses J-Link version v6.54b by default. This version is good enough and the user doesn't need to change for v6.54c (tried on Linux)

The user still needs to set Target Device value as indicated in these instructions.

Comment (kapi-no) [reply doc/nrf/ug_nrf5340.rst]: It works only with Application CPU

File: doc/nrf/ug_nrf5340.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (ru-fu) [reply doc/nrf/ug_nrf5340.rst]: Updated to:

This core is used to communicate with the network.
With regards to the nRF5340 samples, this means that the network core runs the radio stack and real-time procedures.
In the current implementation, the network core runs the controller part of the Bluetooth Low Energy stack.
In general, this core should be used for real-time processing tasks involving low-level protocols and layers.

File: doc/nrf/ug_nrf5340.rst (hunk 12 above)
Comment (ru-fu) [reply doc/nrf/ug_nrf5340.rst]: There's no decision yet on the release of 4.30, so let's leave as is for now.

Comment (ru-fu) [reply doc/nrf/ug_nrf5340.rst]: > 
> 
> It works only with Application CPU

So you cannot program the network core with SES?

File: doc/nrf/ug_nrf5340.rst [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+.. TODO - baso: please check
+
+Communication between the application core and the network core happens through a shared memory area.
+The application core memory is mapped to the network core memory map.
+This means that the network core can access and use the application core memory for shared memory communication.
+
+The cores use Interprocessor Communication (IPC) to notify each other if data is available for the other core to pick up.
+The actual data exchange is handled by Open Asymmetric Multi Processing (OpenAMP).
+
+The IPC driver is implemented in `nrfx`_.
```
[the remark this answers is not in the document]
Comment (ru-fu) [reply doc/nrf/ug_nrf5340.rst]: Replaced this paragraph with:
Zephyr includes the `OpenAMP`_ library, which provides a complete solution for exchanging messages between the cores.
The IPC peripheral is presented to Zephyr as an IPM device (Interprocessor Mailbox).
Th OpenAMP library uses the IPM SHIM layer, which in turn uses the IPC driver in `nrfx`_.

File: doc/nrf/ug_nrf5340.rst [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+   Currently, not everything that is described here is actually supported.
+
+Introduction
+************
+
+nRF5340 is a wireless ultra-low power multicore System on Chip (SoC) with two fully programmable Arm Cortex-M33 processors: a network core and an application core.
+The |NCS| supports Bluetooth Low Energy communication on the nRF5340 SoC.
+
+See the `nRF5340 Product Specification`_ for more information about the nRF5340 SoC.
+nrf5340_dk_nrf5340 in the Zephyr repository gives an overview of the nRF5340 PDK support in Zephyr.
```
Comment (b-gent) [inline doc/nrf/ug_nrf5340.rst]: I understand this is not yet in zephyr?

**Comment by ru-fu** [comment]
Thanks for your comments and additions! I updated the docs.
Please check again - note that we need to merge this PR early tomorrow.

**Comment by b-gent** [comment]
This looks good to me, only one comment.

File: doc/nrf/ug_nrf5340.rst [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+Communication between the application core and the network core happens through a shared memory area.
+The application core memory is mapped to the network core memory map.
+This means that the network core can access and use the application core memory for shared memory communication.
+
+Interprocessor Communication (IPC) is used to indicate to the other core that there is new data available to pick up.
+The actual data exchange is handled by Open Asymmetric Multi Processing (OpenAMP).
+
+Zephyr includes the `OpenAMP`_ library, which provides a complete solution for exchanging messages between the cores.
+The IPC peripheral is presented to Zephyr as an Interprocessor Mailbox (IPM) device.
+Th OpenAMP library uses the IPM SHIM layer, which in turn uses the IPC driver in `nrfx`_.
```
Comment (barsok) [inline doc/nrf/ug_nrf5340.rst]: On my display I see "Th" instead of "The". Could you please check?

File: doc/nrf/ug_nrf5340.rst (hunk 14 above)
Comment (ru-fu) [reply doc/nrf/ug_nrf5340.rst]: Yes, that's why there's no reference. :(
